### PR TITLE
Reuse QgsPostgresConn to prevent deadlock

### PR DESCRIPTION
**Pull request by tomtor**
QgsPostgresProviderConnection::executeSqlPrivate() has a nested
call to itself which may cause a deadlock when the connectionpool is
exhausted:

Two active threads can acquire a Postgres connection at the start of
executeSqlPrivate() and end up both waiting to acquire the second
connection.

## Description

[Replace this with some text explaining the rationale and details about this pull request]

### CODE REVIEW

**Review by m-kuhn — APPROVED** [review]
Good catch.
I guess it's not that easy to come up with a unit test for that

**Comment by tomtor** [comment]
> Good catch.

@m-kuhn Note that I also had an alternative: d8928312a3faf4442c028f4158b8b79f31011d99
but that keeps the connection open. Not sure what is better.

**Comment by elpaso** [comment]
Looks good to me. I should have foreseen this problem.

**Comment by m-kuhn** [comment]
If I'm not wrong, we can have multiple postgres provider connections open in parallel for a long time, therefore keeping it open in there might deplete the connection pool rather quickly. I prefer this approach here.

Sidenote: If you want to go one step further, implementing RAII for the connection (to make `releaseConnection` exception proof) that'd be nice, but it's totally understandable if this is out of scope for this PR.

**Comment by tomtor** [comment]
>  If you want to go one step further, implementing RAII for the connection (to make `releaseConnection` exception proof) that'd be nice, but it's totally understandable if this is out of scope for this PR.

@m-kuhn See https://github.com/qgis/QGIS/pull/40097/commits/1956130d8cabe0d3f12c11ef0d8c6c8cde4ad6b6

as a start. Should be used at more places.

File: src/providers/postgres/qgspostgresconn.h [hunk 1]
```diff
@@ -188,6 +188,15 @@ class QgsPostgresResult
 
 };
 
+class PoolPostgresConn
```
Comment (m-kuhn) [inline src/providers/postgres/qgspostgresconn.h:191]: Can you add documentation that explains the rationale for this class?

Comment (tomtor) [reply src/providers/postgres/qgspostgresconn.h]: done

Comment (tomtor) [reply src/providers/postgres/qgspostgresconn.h]: (and renamed to Pgs...)

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**